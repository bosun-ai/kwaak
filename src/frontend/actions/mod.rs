//! Handles all the actions that can be performed in the frontend based on `UIEvent`
//!
//! NOTE: if we can remove the dependency on app, this could be so much nicer
use std::collections::HashMap;

use copypasta::{ClipboardContext, ClipboardProvider as _};
use strum::{EnumMessage, IntoEnumIterator};

use crate::{chat::Chat, chat_message::ChatMessage, commands::Command, templates::Templates};

use super::{ui_input_command::UserInputCommand, App};

mod diff;
mod github;
mod scroll;

pub use diff::{diff_pull, diff_show};
pub use github::github_issue;
pub use scroll::{scroll_down, scroll_end, scroll_up};

pub fn delete_chat(app: &mut App) {
    let uuid = app.current_chat_uuid;
    app.dispatch_command(uuid, Command::StopAgent);
    // Remove the chat with the given UUID
    app.chats.retain(|chat| chat.uuid != uuid);

    if app.chats.is_empty() {
        app.add_chat(Chat::default());
        app.chats_state.select(Some(0));
        app.add_chat_message(
            app.current_chat_uuid,
            ChatMessage::new_system("Nice, you managed to delete the last chat!"),
        );
    } else {
        app.next_chat();
    }
}

pub fn copy_last_message(app: &mut App) {
    let Some(last_message) = app
        .current_chat()
        .and_then(|c| {
            c.messages
                .iter()
                .filter(|m| {
                    (m.role().is_assistant() || m.role().is_user()) && !m.content().is_empty()
                })
                .next_back()
        })
        .map(ChatMessage::content)
    else {
        app.add_chat_message(
            app.current_chat_uuid,
            ChatMessage::new_system("No message to copy"),
        );
        return;
    }; // Replace with actual retrieval of the last message
       //
    if let Err(e) =
        ClipboardContext::new().and_then(|mut ctx| ctx.set_contents(last_message.to_string()))
    {
        tracing::error!("Error copying last message to clipboard {e:#}");
        return;
    }
    app.add_chat_message(
        app.current_chat_uuid,
        ChatMessage::new_system("Copied last message to clipboard"),
    );
}

pub fn help(app: &mut App) {
    let message = help_message();

    app.add_chat_message(app.current_chat_uuid, ChatMessage::new_system(message));
}

fn help_message() -> String {
    let input_commands_with_description = UserInputCommand::iter()
        .map(|i| (i.to_string(), i.get_documentation()))
        .collect::<HashMap<_, _>>();

    let mut context = tera::Context::new();
    context.insert("commands", &input_commands_with_description);

    Templates::render("help.md", &context).expect("failed to render help message; nice")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_help_message() {
        let message = help_message();
        insta::assert_snapshot!(message);
    }
}
