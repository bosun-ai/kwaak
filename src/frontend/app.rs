use anyhow::Result;
use indoc::indoc;
use std::io;
use std::time::Duration;
use strum::IntoEnumIterator as _;
use text::{ToLine as _, ToSpan};
use tui_logger::TuiWidgetState;
use tui_textarea::TextArea;
use uuid::Uuid;

use ratatui::{
    prelude::*,
    widgets::{Block, Borders, ListState, Padding, Paragraph, Tabs},
};

use crossterm::event::{self, KeyCode, KeyEvent};

use tokio::sync::mpsc;
use tokio::task;

use crate::{
    chat::{Chat, ChatState},
    chat_message::ChatMessage,
    commands::Command,
};

use super::{chat_mode, logs_mode, UIEvent, UserInputCommand};

const TICK_RATE: u64 = 250;
const HEADER: &str = include_str!("ascii_logo");

/// Handles user and TUI interaction
pub struct App<'a> {
    // /// The chat input
    // pub input: String,
    pub text_input: TextArea<'a>,

    /// All known chats
    pub chats: Vec<Chat>,

    /// UUID of the current chat
    pub current_chat: uuid::Uuid,

    /// Holds the sender of UI events for later cloning if needed
    pub ui_tx: mpsc::UnboundedSender<UIEvent>,

    /// Receives UI events (key presses, commands, etc)
    pub ui_rx: mpsc::UnboundedReceiver<UIEvent>,

    /// Sends commands to the backend
    pub command_tx: Option<mpsc::UnboundedSender<Command>>,

    /// Mode the app is in, manages the which layout is rendered and if it should quit
    pub mode: AppMode,

    /// Tracks the current selected state in the UI
    pub chats_state: ListState,

    /// Tab names
    pub tab_names: Vec<&'static str>,

    /// Index of selected tab
    pub selected_tab: usize,

    /// States when viewing logs
    pub log_state: TuiWidgetState,
}

#[derive(Debug, Clone, Copy, Default, PartialEq)]
pub enum AppMode {
    #[default]
    Chat,
    Logs,
    Quit,
}

impl AppMode {
    fn on_key(self, app: &mut App, key: KeyEvent) {
        match self {
            AppMode::Chat => chat_mode::on_key(app, key),
            AppMode::Logs => logs_mode::on_key(app, key),
            AppMode::Quit => (),
        }
    }

    fn ui(self, f: &mut ratatui::Frame, area: Rect, app: &mut App) {
        match self {
            AppMode::Chat => chat_mode::ui(f, area, app),
            AppMode::Logs => logs_mode::ui(f, area, app),
            AppMode::Quit => (),
        }
    }

    fn tab_index(self) -> Option<usize> {
        match self {
            AppMode::Chat => Some(0),
            AppMode::Logs => Some(1),
            AppMode::Quit => None,
        }
    }

    fn from_index(index: usize) -> Option<Self> {
        match index {
            0 => Some(AppMode::Chat),
            1 => Some(AppMode::Logs),
            _ => None,
        }
    }
}

impl Default for App<'_> {
    fn default() -> Self {
        let (ui_tx, ui_rx) = mpsc::unbounded_channel();

        let chat = Chat {
            name: "Chat #1".to_string(),
            ..Chat::default()
        };

        Self {
            text_input: new_text_area(),
            current_chat: chat.uuid,
            chats: vec![chat],
            ui_tx,
            ui_rx,
            command_tx: None,
            mode: AppMode::default(),
            chats_state: ListState::default().with_selected(Some(0)),
            tab_names: vec!["[F1] Chats", "[F2] Logs"],
            log_state: TuiWidgetState::new()
                .set_default_display_level(log::LevelFilter::Off)
                .set_level_for_target("kwaak", log::LevelFilter::Info)
                .set_level_for_target("swiftide", log::LevelFilter::Info),
            selected_tab: 0,
        }
    }
}

fn new_text_area() -> TextArea<'static> {
    let mut text_area = TextArea::default();

    text_area.set_placeholder_text("Send a message to an agent ...");
    text_area.set_placeholder_style(Style::default().fg(Color::Gray));
    text_area.set_cursor_line_style(Style::reset());

    text_area
}

impl App<'_> {
    async fn recv_messages(&mut self) -> Option<UIEvent> {
        self.ui_rx.recv().await
    }

    #[allow(clippy::unused_self)]
    pub fn supported_commands(&self) -> Vec<UserInputCommand> {
        UserInputCommand::iter().collect()
    }

    pub fn send_ui_event(&self, msg: impl Into<UIEvent>) {
        let event = msg.into();
        tracing::debug!("Sending ui event {event}");
        if let Err(err) = self.ui_tx.send(event) {
            tracing::error!("Failed to send ui event {err}");
        }
    }

    pub fn reset_text_input(&mut self) {
        self.text_input = new_text_area();
    }

    fn on_key(&mut self, key: KeyEvent) {
        // Always quit on ctrl c
        if key.modifiers == crossterm::event::KeyModifiers::CONTROL
            && key.code == KeyCode::Char('c')
        {
            tracing::warn!("Ctrl-C pressed, quitting");
            return self.send_ui_event(UIEvent::Quit);
        }

        if let KeyCode::F(index) = key.code {
            let index = index - 1;
            if let Some(mode) = AppMode::from_index(index as usize) {
                return self.change_mode(mode);
            }
        }

        self.mode.on_key(self, key);
    }

    #[tracing::instrument(skip(self))]
    pub fn dispatch_command(&mut self, cmd: &Command) {
        self.current_chat_mut().transition(ChatState::Loading);

        self.command_tx
            .as_ref()
            .expect("Command tx not set")
            .send(cmd.clone())
            .expect("Failed to dispatch command");
    }

    fn add_chat_message(&mut self, message: ChatMessage) {
        let chat = self.find_chat_mut(message.uuid().unwrap_or(self.current_chat));
        chat.add_message(message);
    }

    #[tracing::instrument(skip_all)]
    pub async fn run<B: ratatui::backend::Backend>(
        &mut self,
        terminal: &mut Terminal<B>,
    ) -> io::Result<()> {
        let handle = task::spawn(poll_ui_events(self.ui_tx.clone()));

        loop {
            // Draw the UI
            terminal.draw(|f| {
                let base_area = self.draw_base_ui(f);

                self.mode.ui(f, base_area, self);
            })?;

            if self.mode == AppMode::Quit {
                break;
            }

            // Handle events
            if let Some(event) = self.recv_messages().await {
                if !matches!(event, UIEvent::Tick | UIEvent::Input(_)) {
                    tracing::debug!("Received ui event: {:?}", event);
                }
                match event {
                    UIEvent::Input(key) => {
                        self.on_key(key);
                    }
                    UIEvent::Tick => {
                        // Handle periodic tasks if necessary
                    }
                    UIEvent::CommandDone(uuid) => {
                        self.find_chat_mut(uuid).transition(ChatState::Ready);
                    }
                    UIEvent::AgentActivity(uuid, activity) => {
                        self.find_chat_mut(uuid)
                            .transition(ChatState::LoadingWithMessage(activity));
                    }
                    UIEvent::ChatMessage(message) => {
                        self.add_chat_message(message);
                    }
                    UIEvent::NewChat => {
                        self.add_chat(Chat::default());
                    }
                    UIEvent::NextChat => self.next_chat(),
                    UIEvent::ChangeMode(mode) => self.change_mode(mode),
                    UIEvent::Quit => {
                        tracing::warn!("UI received quit event, quitting");

                        self.dispatch_command(&Command::Quit {
                            uuid: self.current_chat,
                        });
                        self.change_mode(AppMode::Quit);
                    }
                }
            }
        }

        tracing::warn!("Quitting frontend");

        handle.abort();

        Ok(())
    }

    fn find_chat_mut(&mut self, uuid: Uuid) -> &mut Chat {
        self.chats
            .iter_mut()
            .find(|chat| chat.uuid == uuid)
            .unwrap_or_else(|| panic!("Could not find chat for {uuid}"))
    }

    fn find_chat(&self, uuid: Uuid) -> &Chat {
        self.chats
            .iter()
            .find(|chat| chat.uuid == uuid)
            .unwrap_or_else(|| panic!("Could not find chat for {uuid}"))
    }

    pub(crate) fn current_chat(&self) -> &Chat {
        self.find_chat(self.current_chat)
    }

    pub(crate) fn current_chat_mut(&mut self) -> &mut Chat {
        self.find_chat_mut(self.current_chat)
    }

    fn add_chat(&mut self, mut new_chat: Chat) {
        new_chat.name = format!("Chat #{}", self.chats.len() + 1);

        self.current_chat = new_chat.uuid;
        self.chats.push(new_chat);
        self.chats_state.select_last();
    }

    fn next_chat(&mut self) {
        #[allow(clippy::skip_while_next)]
        let Some(next_idx) = self
            .chats
            .iter()
            .position(|chat| chat.uuid == self.current_chat)
            .map(|idx| idx + 1)
        else {
            assert!(
                !cfg!(debug_assertions),
                "Could not find current chat in chats"
            );

            return;
        };

        if let Some(chat) = self.chats.get(next_idx) {
            self.chats_state.select(Some(next_idx));
            self.current_chat = chat.uuid;
        } else {
            self.chats_state.select(Some(0));
            self.current_chat = self.chats[0].uuid;
        }
    }

    fn draw_base_ui(&self, f: &mut Frame) -> Rect {
        let [top_area, main_area] =
            Layout::vertical([Constraint::Length(6), Constraint::Min(0)]).areas(f.area());

        // Hardcoded tabs length for now to right align
        let [header_area, tabs_area] =
            Layout::horizontal([Constraint::Fill(1), Constraint::Length(24)]).areas(top_area);

        Tabs::new(self.tab_names.iter().copied())
            .block(
                Block::default()
                    .borders(Borders::BOTTOM)
                    .padding(Padding::top(top_area.height - 2)),
            )
            .highlight_style(Style::default().add_modifier(Modifier::REVERSED))
            .select(self.selected_tab)
            .render(tabs_area, f.buffer_mut());

        Paragraph::new(HEADER)
            .block(
                Block::default()
                    .borders(Borders::BOTTOM)
                    .padding(Padding::new(1, 0, 1, 0)),
            )
            .render(header_area, f.buffer_mut());

        main_area
    }

    fn change_mode(&mut self, mode: AppMode) {
        self.mode = mode;
        if let Some(tab_index) = mode.tab_index() {
            self.selected_tab = tab_index;
        }
    }
}

#[allow(clippy::unused_async)]
async fn poll_ui_events(ui_tx: mpsc::UnboundedSender<UIEvent>) -> Result<()> {
    loop {
        // Poll for input events
        if event::poll(Duration::from_millis(TICK_RATE))? {
            if let crossterm::event::Event::Key(key) = event::read()? {
                let _ = ui_tx.send(UIEvent::Input(key));
            }
        }
        // Send a tick event, ignore if the receiver is gone
        let _ = ui_tx.send(UIEvent::Tick);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_last_or_first_chat() {
        let mut app = App::default();
        let chat = Chat::default();
        let first_uuid = app.current_chat;
        let second_uuid = chat.uuid;

        // Starts with first
        assert_eq!(app.current_chat, first_uuid);

        app.add_chat(chat);
        assert_eq!(app.current_chat, second_uuid);

        app.next_chat();

        assert_eq!(app.current_chat, first_uuid);

        app.next_chat();
        assert_eq!(app.current_chat, second_uuid);
    }

    #[test]
    fn test_app_mode_from_index() {
        assert_eq!(AppMode::from_index(0), Some(AppMode::Chat));
        assert_eq!(AppMode::from_index(1), Some(AppMode::Logs));
        assert_eq!(AppMode::from_index(2), None);
    }

    #[test]
    fn test_app_mode_tab_index() {
        assert_eq!(AppMode::Chat.tab_index(), Some(0));
        assert_eq!(AppMode::Logs.tab_index(), Some(1));
        assert_eq!(AppMode::Quit.tab_index(), None);
    }

    #[test]
    fn test_change_mode() {
        let mut app = App::default();
        assert_eq!(app.mode, AppMode::Chat);
        assert_eq!(app.selected_tab, 0);

        app.change_mode(AppMode::Logs);
        assert_eq!(app.mode, AppMode::Logs);
        assert_eq!(app.selected_tab, 1);

        app.change_mode(AppMode::Quit);
        assert_eq!(app.mode, AppMode::Quit);
        assert_eq!(app.selected_tab, 1);
    }
}
