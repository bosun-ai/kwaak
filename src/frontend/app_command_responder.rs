use std::sync::Arc;

use async_trait::async_trait;
use tokio::sync::mpsc;
use tokio_util::task::AbortOnDropHandle;
use uuid::Uuid;

use crate::{
    chat_message::ChatMessage,
    commands::{CommandResponse, Responder},
};

use super::ui_event::UIEvent;

/// Handles responses from commands application wide
///
/// Basically converts command responses into UI events
/// The responder is send with commands so that the backend has a way to communicate with the
/// frontend, without knowing about the frontend
///
/// When dispatching commands, a responder is created that wraps each response with the chat id
#[derive(Debug)]
pub struct AppCommandResponder {
    // ui_tx: mpsc::UnboundedSender<UIEvent>,
    tx: mpsc::UnboundedSender<ResponseWithChatId>,
    _handle: AbortOnDropHandle<()>,
}

#[derive(Debug)]
struct ResponseWithChatId(Uuid, CommandResponse);

#[derive(Debug, Clone)]
pub struct AppCommandResponderForChatId {
    inner: mpsc::UnboundedSender<ResponseWithChatId>,
    uuid: uuid::Uuid,
}

impl AppCommandResponder {
    pub fn spawn_for(ui_tx: mpsc::UnboundedSender<UIEvent>) -> AppCommandResponder {
        tracing::info!("Initializing app command responder");
        let (tx, mut rx) = mpsc::unbounded_channel::<ResponseWithChatId>();
        let handle = tokio::spawn(async move {
            while let Some(response) = rx.recv().await {
                tracing::debug!("[RESPONDER] Received response: {:?}", response);
                let chat_id = response.0;
                let ui_event = match response.1 {
                    CommandResponse::Chat(msg) => UIEvent::ChatMessage(chat_id, msg.into()),
                    CommandResponse::Activity(state) => UIEvent::ActivityUpdate(chat_id, state),
                    CommandResponse::RenameChat(name) => UIEvent::RenameChat(chat_id, name),
                    CommandResponse::RenameBranch(name) => UIEvent::RenameBranch(chat_id, name),
                    CommandResponse::Completed => UIEvent::CommandDone(chat_id),
                    CommandResponse::BackendMessage(msg) => {
                        UIEvent::ChatMessage(chat_id, ChatMessage::new_system(&msg))
                    }
                };

                if let Err(err) = ui_tx.send(ui_event) {
                    tracing::error!("Failed to send response to ui: {:#}", err);
                }
            }
            tracing::info!("App command responder shutting down");
        });

        // Create the app responder, spawn a task to handle responses, the once cell returns a
        // clone of the responder without the rx
        AppCommandResponder {
            tx,
            _handle: AbortOnDropHandle::new(handle),
        }
    }

    #[must_use]
    pub fn for_chat_id(&self, uuid: Uuid) -> Arc<dyn Responder> {
        Arc::new(AppCommandResponderForChatId {
            inner: self.tx.clone(),
            uuid,
        }) as Arc<dyn Responder>
    }
}

#[async_trait]
impl Responder for AppCommandResponderForChatId {
    async fn send(&self, response: CommandResponse) {
        tracing::debug!("[RESPONDER SENDER] Sending response: {:?}", response);
        let response = ResponseWithChatId(self.uuid, response);
        if let Err(err) = self.inner.send(response) {
            tracing::error!("Failed to send response for command: {:?}", err);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tokio::sync::mpsc;
    use uuid::Uuid;

    const TEST_UUID: Uuid = Uuid::from_u128(0x1234_5678_90ab_cdef_1234_5678_90ab_cdef);

    #[tokio::test]
    async fn test_app_responder() {
        let (ui_tx, mut ui_rx) = mpsc::unbounded_channel();
        let app = AppCommandResponder::spawn_for(ui_tx);

        let responder = app.for_chat_id(TEST_UUID);

        responder.system_message("Test message").await;

        let Some(ui_event) = ui_rx.recv().await else {
            panic!("No UI event received");
        };

        match ui_event {
            UIEvent::ChatMessage(received_uuid, received_message) => {
                assert_eq!(received_uuid, TEST_UUID);
                assert_eq!(received_message.content(), "Test message");
                assert!(received_message.role().is_system());
            }
            _ => panic!("Unexpected UI event received"),
        }

        responder.send(CommandResponse::Completed).await;

        // Verify the UI event is received
        if let Some(ui_event) = ui_rx.recv().await {
            match ui_event {
                UIEvent::CommandDone(received_uuid) => assert_eq!(received_uuid, TEST_UUID),
                _ => panic!("Unexpected UI event received"),
            }
        } else {
            panic!("No UI event received");
        }
    }
}
