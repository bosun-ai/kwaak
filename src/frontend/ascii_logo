           _                     _   
   (@_    | |__ _ _ _  ___  ___ | |__
\\\_\     | / /| | | |<_> |<_> || / /
<____)    |_\_\|__/_/ <___|<___||_\_\
                                      



