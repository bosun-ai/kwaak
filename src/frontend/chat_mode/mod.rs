mod chat_list_widget;
mod chat_messages_widget;
mod help_section_widget;
mod input_bar_widget;
mod message_formatting;
mod on_key;
mod ui;

pub use on_key::on_key;
pub use ui::ui;
