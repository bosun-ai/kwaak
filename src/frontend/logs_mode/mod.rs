mod on_key;
mod ui;

pub use on_key::on_key;
pub use ui::ui;
