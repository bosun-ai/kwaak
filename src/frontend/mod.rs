mod app;
mod splash;
mod ui_event;
mod ui_input_command;

mod actions;
mod app_command_responder;
/// Different frontend ui modes
mod chat_mode;
mod logs_mode;

/// Let's be very strict about what to export
/// to avoid coupling frontend and the rest
pub use app::App;

#[cfg(debug_assertions)]
pub use app_command_responder::AppCommandResponder;
#[cfg(debug_assertions)]
pub use chat_mode::ui;
#[cfg(debug_assertions)]
pub use ui_event::UIEvent;
#[cfg(debug_assertions)]
pub use ui_input_command::{DiffVariant, UserInputCommand};
