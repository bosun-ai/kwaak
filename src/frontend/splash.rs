use std::borrow::Cow;

use layout::Flex;
use ratatui::{
    Frame,
    widgets::{Block, Padding},
};
use ratatui_splash_screen::{SplashConfig, SplashScreen};

// use std::error::Error;
// use std::io::stdout;
// use std::time::Duration;
//
use ratatui::prelude::*;
use text::ToSpan as _;
// use ratatui_splash_screen::{SplashConfig, SplashScreen, SplashError};
//
static SPLASH_CONFIG: SplashConfig = SplashConfig {
    image_data: include_bytes!("../../images/logo_term_nobg.png"),
    sha256sum: None,
    render_steps: 6,
    use_colors: true,
};

pub struct Splash<'m> {
    splash_screen: SplashScreen,
    message: Cow<'m, str>, // Might get tens of thousands of updates (or more), so let's avoid unnecessary allocations
}

impl Default for Splash<'_> {
    fn default() -> Self {
        let splash_screen = SplashScreen::new(SPLASH_CONFIG).unwrap();
        Splash {
            splash_screen,
            message: "".into(),
        }
    }
}

// Splash renders in the center of the screen an image with a loading bar
// Image is 50x50 characters, below a line with the text "Indexing your code ..."
impl<'m> Splash<'m> {
    pub fn render(&mut self, f: &mut Frame) {
        let splash_area = center(
            f.area(),
            Constraint::Percentage(48),
            Constraint::Percentage(67),
        );
        let [image_area, text_area] =
            Layout::vertical([Constraint::Fill(1), Constraint::Length(2)]).areas(splash_area);

        f.render_widget(&mut self.splash_screen, image_area);

        #[allow(clippy::cast_possible_truncation)]
        let left_padding = (text_area.width - self.message.len() as u16) / 2;
        let block = Block::default().padding(Padding::new(left_padding, 0, 1, 0));
        let throbber = throbber_widgets_tui::Throbber::default().label(self.message.to_span());

        f.render_widget(throbber, block.inner(text_area));
        f.render_widget(block, text_area);
    }

    pub fn is_rendered(&self) -> bool {
        self.splash_screen.is_rendered()
    }

    pub fn set_message<T>(&mut self, message: T)
    where
        T: Into<Cow<'m, str>>,
    {
        self.message = message.into();
    }
}

fn center(area: Rect, horizontal: Constraint, vertical: Constraint) -> Rect {
    let [area] = Layout::horizontal([horizontal])
        .flex(Flex::Center)
        .areas(area);
    let [area] = Layout::vertical([vertical]).flex(Flex::Center).areas(area);
    area
}
