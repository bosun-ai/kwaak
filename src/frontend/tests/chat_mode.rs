use insta::assert_snapshot;
use ratatui::{backend::TestBackend, Terminal};
use uuid::Uuid;

use crate::frontend::{chat_mode, App};

/// Simple snapshots for now so we can refactor later.
#[test]
fn test_render_app() {
    let mut app = App::default();
    let fixed_uuid = Uuid::parse_str("936DA01F9ABD4d9d80C702AF85C822A8").unwrap();
    app.current_chat = fixed_uuid;
    let mut terminal = Terminal::new(TestBackend::new(160, 40)).unwrap();

    terminal
        .draw(|f| chat_mode::ui(f, f.area(), &mut app))
        .unwrap();
    assert_snapshot!(terminal.backend());
}
