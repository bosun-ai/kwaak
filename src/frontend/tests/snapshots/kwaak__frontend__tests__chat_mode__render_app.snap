---
source: src/frontend/tests/chat_mode.rs
expression: terminal.backend()
---
"                                                                                                                                ─────────────Chats─────────────┐"
"                                                                                                                                 󰍩  Chat #1                    │"
"                                                                                                                                                               │"
"                                                                                                                                                               │"
"                                                                                                                                                               │"
"                                                                                                                                                               │"
"                                                                                                                                                               │"
"                                                                                                                                                               │"
"                                                                                                                                                               │"
"                                                                                                                                                               │"
"                                                                                                                                                               │"
"                                                                                                                                                               │"
"                                                                                                                                                               │"
"                                                                                                                                                               │"
"                                                                                                                                                               │"
"                                                                                                                                                               │"
"                                                                                                                                                               │"
"                                                                                                                                                               │"
"                                                                                                                                ─────────Chat commands─────────┤"
"                                                                                                                                                               │"
"                                                                                                                                 /quit                         │"
"                                                                                                                                 /show_config                  │"
"                                                                                                                                 /index_repository             │"
"                                                                                                                                 /next_chat                    │"
"                                                                                                                                 /new_chat                     │"
"                                                                                                                                 /delete_chat                  │"
"                                                                                                                                 /copy                         │"
"                                                                                                                                                               │"
"                                                                                                                                ──────────Keybindings──────────┤"
"                                                                                                                                                               │"
"├──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤ Page Up/Down - Scroll         │"
"│  Send a message to an agent ...                                                                                              │ End - Scroll to end           │"
"│                                                                                                                              │ ^s - Send message             │"
"│                                                                                                                              │ ^x - Stop agent               │"
"│                                                                                                                              │ ^n - New chat                 │"
"│                                                                                                                              │ ^q - Quit                     │"
"│                                                                                                                              │                               │"
"└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴───────────────────────────────┘"
"                                                                                                                     kwaak/936da01f-9abd-4d9d-80c7-02af85c822a8 "
"                                                                                                                                                                "
